src/mds/MDCache.cc

```diff
@@ -13411,6 +13473,189 @@ void MDCache::clear_dirty_bits_for_stray(CInode* diri) {
   }
 }
 
+void MDCache::dispatch_quiesce_subvolume_inode(const MDRequestRef& mdr)
+{
+  if (mdr->killed) {
+    dout(20) << __func__ << " " << *mdr << " not dispatching killed " << *mdr << dendl;
+    return;
+  } else if (mdr->internal_op_finish == nullptr) {
+    dout(20) << __func__ << " " << *mdr << " already finished quiesce" << dendl;
+    return;
+  }
+
+  CInode *in = get_inode(mdr->get_filepath().get_ino());
+  if (in == nullptr) {
+    mds->server->respond_to_request(mdr, -ENOENT);
+    return;
+  }
+
+  dout(20) << __func__ << " " << *mdr << " locking " << *in << dendl;
+
+  if (in->is_auth()) {
+    MutationImpl::LockOpVec lov;
+    lov.add_rdlock(&in->authlock);
+    lov.add_rdlock(&in->dirfragtreelock);
+    if (in->is_file()) {
+      /* Because files are treated specially allowing multiple reader/writers,
+       * we need an xlock here to recall all write caps. This unfortunately
+       * means there can be no readers.
+       */
+      lov.add_xlock(&in->filelock);
```

Comment:
maybe we could demote this to a readlock once we quiesce successfully, thus allowing readers?

Reply:
No, the MDS remembers the wanted caps for clients and would reissue write caps.